KDE: Add KMyMoney

aszlig opened the pull request:
This adds [KMyMoney](https://kmymoney.org/), a finance manager for KDE plus a few required dependencies.

I ran the upstream test suite as well as the following manual tests:

  * Basic startup
  * Completing the wizard
  * Add some test transactions
  * GPG encryption
  * Generation of charts and reports
  * Rough check whether OFX integration lists supported financial institutions.
  * Small check of AqBanking implementation, whether accounts and users can be configured, but didn't test actual connectivity with a financial institution.
  * Check of Weboob integration with a test PayPal backend, however also just with a dummy account and without actually connecting to PayPal.

I also added @7c6f434c to the reviewers, because he's the maintainer of `mpir` where I had to enable C++ support (see 6134887bef5357ccda4d81aef78182217df4c3ed for details).

Note that I didn't add myself as a maintainer because I'm not personally using KMyMoney but just packaged it for someone else. I hope this is useful for other people, so that maybe someday™ there will be a proper maintainer.

aszlig committed:
mpir: Build with C++ support

This is needed by Alkimia which I'm going to package soon as part of
KMyMoney.

I enabled this by default, because it doesn't increase the closure size
a whole lot (only around 150 KiB):

$ nix-store -q --size /nix/store/...old...-mpir-3.0.0
1223248
$ nix-store -q --size /nix/store/...new...-mpir-3.0.0
1377136

Introducing an option for enabling/disabling this is not worth it,
because it doesn't conflict with anything and the size increase is the
only drawback and we can still make it configurable if we want someday.

Signed-off-by: aszlig <aszlig@nix.build>
Cc: @7c6f434c

aszlig committed:
kdewebkit: Add package expression

This is one of the requirements of KMyMoney and packaging is quite
straightforward with no unexpected traps.

Signed-off-by: aszlig <aszlig@nix.build>
Cc: @ttuegel

aszlig committed:
alkimia: init at 7.0.1

This is used by KMyMoney and also the reason why I needed to add C++
support to MPIR in the parent commit.

The reason why I didn't add myself as a maintainer is because I'm not
personally using KMyMoney and thus Alkimia.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
kmymoney: init at 5.0.1

I tried to use -DENABLE_SQLCIPHER and also passed the right directories
to the Qt 5 source of the QSQLiteDriver but CMake then failed to run
qt4_automoc, by which I'd imply that SQLCipher is not maintained
anymore (after all KMyMoney using qgpgme as well, which doesn't require
sources).

Another odd thing is that CMake reports that the weboob plugin is
disabled, but after inspecting it turns out that the reporting is just
wrong. This is already fixed upstream but not yet released in
KDE/kmymoney@8b086cf921879770971e9331f005339bfacdc84c.

In addition of running the upstream test suite I have manually tested a
few things in a VM by using the following Nix expression:

  (import <nixpkgs/nixos> {
    configuration = { pkgs, ... }: {
      environment.systemPackages = with pkgs; [
        kmymoney aqbanking gwenhywfar libchipcard python2Packages.weboob
        kgpg
      ];
      users.users.test.isNormalUser = true;
      virtualisation.diskSize = 4096;
      virtualisation.memorySize = 2048;
      services.xserver = {
        enable = true;
        inherit ((import <nixpkgs/nixos> {}).config.services.xserver) layout;
        displayManager.sddm.enable = true;
        displayManager.sddm.autoLogin.enable = true;
        displayManager.sddm.autoLogin.user = "test";
        desktopManager.default = "plasma5";
        desktopManager.plasma5.enable = true;
      };
    };
  }).vm

The things I have tested in particular are:

  * Basic startup
  * Completing the wizard
  * Add some test transactions
  * GPG encryption
  * Generation of charts and reports
  * Rough check whether OFX integration lists supported financial
    institutions.
  * Small check of AqBanking implementation, whether accounts and users
    can be configured, but didn't test actual connectivity with a
    financial institution.
  * Check of Weboob integration with a test PayPal backend, however also
    just with a dummy account and without actually connecting to PayPal.

One of the upstream tests "reports-chart-test" seems to fail even though
generating charts and reports are working when testing manually. It also
seems that this is the case on other distributions, for example Gentoo
has disabled that test as well:

https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=5169cec68fa6fd67841

Note that I didn't add myself as a maintainer because I'm not personally
using KMyMoney but just packaged it for someone else. I hope this is
useful for other people, so that maybe someday[TM] there will be a
proper maintainer.

Signed-off-by: aszlig <aszlig@nix.build>
Cc: @ttuegel

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39647&attempt_id=b2bbc905-f344-47c7-88bf-bf2eb3358e36)

The following builds were skipped because they don't evaluate on x86_64-darwin: kmymoney, mpir

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39647&attempt_id=056acdb9-a02f-4ad2-9bc1-a913d03656ee)

Attempted: kmymoney, mpir

<details><summary>Partial log (click to expand)</summary><p>

```
      Start 37: kmymoney-kmymoneyutils-test
37/37 Test #37: kmymoney-kmymoneyutils-test ......................   Passed    0.05 sec

100% tests passed, 0 tests failed out of 37

Total Test time (real) = 108.49 sec
postPatchMkspecs
postPatchMkspecs
/nix/store/g6bi304qx5bmpk0am13p6yylv1rgrgm1-kmymoney-5.0.1
/nix/store/7wwrbw3r5qi27ib6acm2x9hwg6xkwi7m-mpir-3.0.0
```
</p></details>

7c6f434c commented:
1. As for MPIR — seems fine, as far as I remember it is for eventual use in Sage, size doesn't matter that much.

2. Maybe copy-paste the upstream «Alkimia is the infrastructure for common storage and business logic that will be used by all financial applications in KDE. The target is to share financial related information over application bounderies. » as `longDescription` for Alkimia? The current `description` makes me want to look up the homepage to find out what it actually is (then I do, and despair).

3. We need to do something about the notion of maintenance, in the that the initial definition are people who can be asked because they know what is going on in the expression, and then everyone has different interpretations.

aszlig committed:
alkimia: Add meta.longDescription

@7c6f434c wrote in [1]:

   Maybe copy-paste the upstream "Alkimia is the infrastructure for
   common storage and business logic that will be used by all financial
   applications in KDE. The target is to share financial related
   information over application bounderies." as longDescription for
   Alkimia? The current description makes me want to look up the
   homepage to find out what it actually is (then I do, and despair).

[1]: https://github.com/NixOS/nixpkgs/pull/39647#issuecomment-385169261

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@7c6f434c: Point 2 fixed, as for point 3, to me `meta.maintainers` means `I will watch this package and continuously make improvements and updates.`, after all there is `git shortlog` to see who's made the most changes. So I'd guess we don't have a general consensus on that, right?

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39647&attempt_id=1cf83a4b-6fc5-4f2d-8f6a-8d75c4d93b79)

Attempted: kmymoney, mpir

<details><summary>Partial log (click to expand)</summary><p>

```
[4734/17529] CXX obj/third_party/icu/bundled_icui18n/visibledigits.o
[4735/17529] CXX obj/third_party/icu/bundled_icui18n/regexcmp.o
[4736/17529] CXX obj/services/catalog/lib/catalog.o
[4737/17529] CXX obj/third_party/icu/bundled_icuuc/ucnvscsu.o
[4738/17529] CXX obj/third_party/icu/bundled_icuuc/unames.o
building of '/nix/store/iq5c59ld0j6zff4vs7a72dri54h2vidb-qtwebengine-5.10.1.drv' timed out after 3600 seconds
cannot build derivation '/nix/store/yl4r9978d9pb5bgwfqydb8s9k8jj7052-python2.7-PyQt-5.10.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/7c2cyxv43cvl4272dm365vqgn15mvq9k-python2.7-weboob-1.3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0fsw05f97d1wj6km6fba68xswbg1ag5z-kmymoney-5.0.1.drv': 1 dependencies couldn't be built
[31;1merror:[0m build of '/nix/store/0fsw05f97d1wj6km6fba68xswbg1ag5z-kmymoney-5.0.1.drv' failed
```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39647&attempt_id=b3258712-b6f6-4a75-9272-12f8bea13b73)

The following builds were skipped because they don't evaluate on x86_64-darwin: kmymoney, mpir

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39647&attempt_id=ca6fafc5-bbd0-4c3d-885c-94eac443411e)

Attempted: kmymoney, mpir

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/g6bi304qx5bmpk0am13p6yylv1rgrgm1-kmymoney-5.0.1
/nix/store/7wwrbw3r5qi27ib6acm2x9hwg6xkwi7m-mpir-3.0.0
```
</p></details>

7c6f434c commented:
Well, it doesn't define what it does say. The initial mailing list discussion ended with a rather minimal approach, and no consensus to make an explicit changed has since been reached…

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39647&attempt_id=e8d8bcf8-4d93-47dc-8452-e0311b5bff6e)

Attempted: kmymoney, mpir

<details><summary>Partial log (click to expand)</summary><p>

```
[4622/17529] CXX obj/third_party/icu/bundled_icui18n/repattrn.o
[4623/17529] CXX obj/third_party/icu/bundled_icui18n/sharedbreakiterator.o
[4624/17529] CXX obj/third_party/icu/bundled_icui18n/collationtailoring.o
[4625/17529] CXX obj/third_party/icu/bundled_icui18n/standardplural.o
[4626/17529] CXX obj/third_party/icu/bundled_icui18n/search.o
building of '/nix/store/iq5c59ld0j6zff4vs7a72dri54h2vidb-qtwebengine-5.10.1.drv' timed out after 3600 seconds
cannot build derivation '/nix/store/yl4r9978d9pb5bgwfqydb8s9k8jj7052-python2.7-PyQt-5.10.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/7c2cyxv43cvl4272dm365vqgn15mvq9k-python2.7-weboob-1.3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0fsw05f97d1wj6km6fba68xswbg1ag5z-kmymoney-5.0.1.drv': 1 dependencies couldn't be built
[31;1merror:[0m build of '/nix/store/0fsw05f97d1wj6km6fba68xswbg1ag5z-kmymoney-5.0.1.drv' failed
```
</p></details>
